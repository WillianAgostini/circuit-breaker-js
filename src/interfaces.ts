export interface Options {
    windowSize?: number;
    failureThresholdPercentage?: number;
    timeout?: number,
    resetTimeout?: number,
    isError?(err: any): boolean;
    autoRenewAbortController?: boolean;
}
