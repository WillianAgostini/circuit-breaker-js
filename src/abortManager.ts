import { randomUUID } from 'crypto';

export class AbortManager {
    readonly abortControllers = new Map<string, AbortController>();

    create() {
        const id = randomUUID();
        const abortController = new AbortController();
        this.abortControllers.set(id, abortController);
        return { id, abortController };
    }

    abort(id: string, reason?: string): boolean {
        const abortController = this.abortControllers.get(id);
        if (!abortController) return false;

        abortController.abort(reason);
        return this.abortControllers.delete(id);
    }

    abortAll(reason?: string) {
        this.abortControllers.forEach(abortController => abortController.abort(reason));
        this.abortControllers.clear();
    }
}
