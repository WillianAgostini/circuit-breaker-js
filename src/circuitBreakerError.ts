export class CircuitBreakerError extends Error { }
