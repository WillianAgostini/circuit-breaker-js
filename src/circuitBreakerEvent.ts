import { EventEmitter } from "node:events";

export class CircuitBreakerEvent extends EventEmitter {

    emit(event: 'open'): boolean;
    emit(event: 'close'): boolean;
    emit(event: 'halfOpen'): boolean;
    emit(event: 'sucess', response: any): boolean;
    emit(event: 'error', err: any): boolean;
    emit(event: string, ...args: any[]): boolean {
      return super.emit(event, ...args);
    }
  
    on(event: 'open', listener: () => void): this;
    on(event: 'close', listener: () => void): this;
    on(event: 'halfOpen', listener: () => void): this;
    on(event: 'sucess', listener: (response: any) => void): this;
    on(event: 'error', listener: (err: any) => void): this;
    on(event: string, listener: (...args: any[]) => void): this {
      return super.on(event, listener);
    }

}