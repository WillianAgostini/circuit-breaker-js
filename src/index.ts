export * from "./circuitBreaker";
export * from "./state";
export * from "./interfaces";


// (async () => {
//     const promise = new Promise(resolve => setTimeout(resolve, 10));

//     const breaker = new CircuitBreaker({ timeout: 1000 });
//     try {
//         await breaker.execute(promise)
//     } catch (error) {
//         console.log(error)
//     }

// })()