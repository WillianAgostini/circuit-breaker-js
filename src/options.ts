export interface Options {
    /** 
     * Time window in milliseconds to track failures and successes.
     * Used to calculate failure rates within the given timeframe.
     * @default 60000
     */
    windowSize?: number;

    /** 
     * Percentage threshold of failures within the windowSize that triggers the circuit to open.
     * @default 5
     */
    failureThresholdPercentage?: number;

    /** 
     * Maximum time in milliseconds allowed for a request before it is considered a failure.
     * @default undefined
     */
    timeout?: number;

    /** 
     * Time in milliseconds the circuit remains open before attempting to transition to half-open.
     * @default undefined
     */
    resetTimeout?: number;

    /** 
     * Custom error-checking function to determine if an error should count as a failure.
     * @default undefined
     */
    isError?: (err: any) => boolean;

    /** 
     * Alternative to failureThresholdPercentage; opens the circuit when failures reach this count within the window.
     * @default 0
     */
    failureThresholdCount?: number;

    /** 
     * Number of allowed retry attempts in the half-open state before opening the circuit again if failures persist.
     * @default 1
     */
    retryAttempts?: number;

    /** 
     * Number of consecutive successes required in the half-open state to fully close the circuit.
     * @default 1
     */
    successThreshold?: number;
}