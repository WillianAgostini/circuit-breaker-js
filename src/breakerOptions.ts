import { Options } from "./options";

export class BreakerOptions {
    windowSize: number;
    failureThresholdPercentage: number;
    timeout?: number;
    resetTimeout?: number;
    isError?: (err: any) => boolean;
    autoRenewAbortController: boolean;
    failureThresholdCount: number;
    successThreshold: number;

    constructor(opts: Options) {
        const {
            failureThresholdPercentage = 5,
            windowSize = 60000,
            timeout,
            resetTimeout,
            isError,
            autoRenewAbortController = false,
            failureThresholdCount = 0,
            successThreshold = 1,
        } = opts;

        this.failureThresholdPercentage = failureThresholdPercentage;
        this.windowSize = windowSize;
        this.timeout = timeout;
        this.resetTimeout = resetTimeout;
        this.isError = isError;
        this.autoRenewAbortController = autoRenewAbortController;
        this.failureThresholdCount = failureThresholdCount;
        this.successThreshold = successThreshold;
    }
}
