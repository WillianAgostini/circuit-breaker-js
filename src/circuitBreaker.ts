import { AbortManager } from "./abortManager";
import { BreakerOptions } from "./breakerOptions";
import { CircuitBreakerError } from "./circuitBreakerError";
import { CircuitBreakerEvent } from "./circuitBreakerEvent";
import { Options } from "./options";
import { State } from "./state";
import { Window } from "./window";

/**
 * Represents a function that executes an asynchronous operation with cancellation support.
 * 
 * @template T The type of the resolved value of the promise.
 */
export interface PromiseFunction<T> {
    /**
     * Executes the operation.
     * 
     * @param signal - An `AbortSignal` that allows the operation to be canceled.
     * If the operation is aborted (e.g., due to a timeout or circuit breaker),
     * the promise will reject with an `AbortError`.
     * 
     * @returns A promise that resolves with a value of type `T` or rejects 
     * if an error occurs or if the operation is aborted.
     */
    (signal: AbortSignal): Promise<T>;
}

export class CircuitBreaker {
    readonly event = new CircuitBreakerEvent();

    readonly #state: State;
    readonly #window: Window;
    readonly #options: BreakerOptions;
    readonly #abortManager = new AbortManager();
    #resetTimeout: NodeJS.Timeout | undefined;

    constructor(opts: Options) {
        this.#options = new BreakerOptions(opts);
        this.#state = new State();
        this.#window = new Window(this.#options.windowSize);
    }

    get successCount() {
        return this.#window.successCount;
    }

    get failureCount() {
        return this.#window.failureCount;
    }

    get totalRequests() {
        return this.#window.totalRequests;
    }

    get failurePercentage() {
        return this.#window.failurePercentage;
    }

    isOpen() {
        return this.#state.isOpen;
    }

    isClosed() {
        return this.#state.isClosed;
    }

    isHalfOpen() {
        return this.#state.isHalfOpen;
    }

    open() {
        this.#state.setOpen();
        this.#startResetTimer();
        this.#window.reset();
        // this.#abortControllers.abortAll();
        this.#state.setAttemptingClose(false);
        this.event.emit('open');
    }

    close() {
        this.#clearResetTimer();
        this.#state.setClosed();
        this.#state.setAttemptingClose(false);
        this.#window.reset();
        this.event.emit('close');
    }

    halfOpen() {
        this.#state.setHalfOpen();
        this.#state.setAttemptingClose(true);
        this.event.emit('halfOpen');
    }

    async execute<T>(promiseFunction: PromiseFunction<T>) {
        this.#ensureIsPromiseFunction(promiseFunction);
        this.#evaluateResetCondition();

        if (this.isOpen()) {
            this.event.emit('reject');
            throw new CircuitBreakerError('Circuit is open');
        }

        if (this.isHalfOpen()) {
            if (!this.#state.canTryClosing()) {
                this.event.emit('reject');
                throw new CircuitBreakerError('Circuit is halfOpen and this executions was rejected');
            }

            this.#state.setAttemptingClose(false);
        }

        const { id, abortController } = this.#abortManager.create();
        try {
            const result = await Promise.race([promiseFunction(abortController.signal), this.#timeoutRejection(abortController.signal)]);
            this.#abortManager.abort(id);
            this.#handleExecutionSuccess(result);
            return result as T;
        } catch (error) {
            this.#abortManager.abort(id);
            this.#handleExecutionError(error);
        }
    }

    #ensureIsPromiseFunction<T>(promiseFn: PromiseFunction<T>) {
        if (typeof promiseFn !== 'function') {
            throw new TypeError(`Invalid argument: Expected a function that returns a Promise. Ensure you are passing a function, not a Promise.`);
        }
    }

    #handleExecutionSuccess(result?: any) {
        if (this.isHalfOpen()) {
            this.#window.recordSuccessOnHalfOpen();
            this.#evaluateCloseCondition();
            this.#state.setAttemptingClose(true);
        }
        this.#window.recordSuccess();
        this.event.emit('success', result);
    }

    #handleExecutionError(error: any) {
        if (this.#options.isError && !this.#options.isError(error)) {
            this.#handleExecutionSuccess();
            throw error;
        }

        if (this.isHalfOpen()) {
            this.#window.recordFailureOnHalfOpen();
            this.#window.resetSuccessOnHalfOpen();
            this.#evaluateOpenCondition();
            this.#state.setAttemptingClose(true);
        }

        if (this.isClosed()) {
            this.#window.recordFailure();
            this.#evaluateResetCondition();
        }

        this.event.emit('error', error);
        throw error;
    }

    #evaluateCloseCondition() {
        if (this.#window.successCountOnHalfOpen >= this.#options.successThreshold) {
            this.close();
        }
    }

    #evaluateOpenCondition() {
        if (this.#window.failureCountOnHalfOpen >= this.#options.retryAttempts) {
            this.open();
        }
    }

    #timeoutRejection(signal: AbortSignal) {
        if (!this.#options.timeout) return;

        return new Promise((_, reject) => {
            const timeoutId = setTimeout(() => {
                clearTimeout(timeoutId);
                reject(new Error('Operation timed out'));
            }, this.#options.timeout);

            signal.addEventListener('abort', () => {
                clearTimeout(timeoutId);
            });
        });
    }

    #evaluateResetCondition() {
        if (this.#window.failurePercentage > this.#options.failureThresholdPercentage) {
            return this.open();
        }

        if (this.#options.failureThresholdCount && this.#window.failureCount >= this.#options.failureThresholdCount) {
            this.open();
        }
    }

    #startResetTimer() {
        if (!this.#options.resetTimeout) return;

        this.#resetTimeout = setTimeout(() => this.halfOpen(), this.#options.resetTimeout);
    }

    #clearResetTimer() {
        clearTimeout(this.#resetTimeout);
    }

    toJSON() {
        return {
            isOpen: this.isOpen(),
            isClosed: this.isClosed(),
            isHalfOpen: this.isHalfOpen(),
            successCount: this.successCount,
            failureCount: this.failureCount,
            totalRequests: this.totalRequests,
            failurePercentage: this.failurePercentage,
            options: this.#options,
        };
    }
}
