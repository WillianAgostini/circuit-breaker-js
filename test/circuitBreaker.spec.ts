import { CircuitBreaker } from '../src/index';

const successPromise = (value: string) => Promise.resolve(value);
const failurePromise = (timeout: number, error: Error) => new Promise((_, reject) => setTimeout(() => reject(error), timeout));
const timeoutPromise = (timeout: number) => new Promise((resolve) => setTimeout(resolve, timeout));

describe('CircuitBreaker', () => {

    describe('execute', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({ timeout: 100, resetTimeout: 5 });
        });

        test('should execute successfully', async () => {
            const promise = () => successPromise('success');
            await expect(breaker.execute(promise)).resolves.toBe('success');
        });

        test('should throw error on timeout', async () => {
            const promise = () => timeoutPromise(110);
            await expect(breaker.execute(promise)).rejects.toThrow('Operation timed out');
        });

        test('should set state to OPEN on failure', async () => {
            const promise = () => failurePromise(5, new Error('failure'));
            await expect(breaker.execute(promise)).rejects.toThrow('failure');
            expect(breaker.isOpen()).toBe(true);
            await timeoutPromise(5);
            expect(breaker.isHalfOpen()).toBe(true);
            await breaker.execute(() => successPromise('success'));
            expect(breaker.isClosed()).toBe(true);
        });

        test('should throw error when circuit is OPEN', async () => {
            breaker.open();
            const promise = () => successPromise('success');
            await expect(breaker.execute(promise)).rejects.toThrow('Circuit is open');
        });

        test('should not sum error when circuit is OPEN', async () => {
            breaker.open();
            const promise = () => successPromise('success');
            await expect(breaker.execute(promise)).rejects.toThrow('Circuit is open');
            expect(breaker.successCount).toBe(0);
            expect(breaker.failureCount).toBe(0);
            expect(breaker.totalRequests).toBe(0);
            expect(breaker.failurePercentage).toBe(0);
        });

        test('should pass 1 execution when circuit is HALF_OPEN', async () => {
            breaker.halfOpen();
            const successMock = jest.fn();
            const rejectMock = jest.fn();
            breaker.event.on('success', successMock);
            breaker.event.on('reject', rejectMock);

            const promise = () => timeoutPromise(1);

            await Promise.allSettled([
                breaker.execute(promise),
                breaker.execute(promise),
                breaker.execute(promise),
            ]);

            expect(successMock).toHaveBeenCalledTimes(1);
            expect(rejectMock).toHaveBeenCalledTimes(2);
        });

        test('should pass 1 execution when circuit is HALF_OPEN', (done) => {
            breaker.halfOpen();
            const successMock = jest.fn();
            const rejectMock = jest.fn();
            const errorMock = jest.fn();
            breaker.event.on('success', successMock);
            breaker.event.on('reject', rejectMock);
            breaker.event.on('error', errorMock);

            const failPromise = () => failurePromise(5, new Error('failure'));
            const promise = () => timeoutPromise(5);

            Promise.allSettled([
                breaker.execute(failPromise),
                breaker.execute(failPromise),
            ]).finally(() => {
                setTimeout(async () => {
                    await breaker.execute(promise);

                    expect(breaker.isClosed()).toBe(true);
                    expect(successMock).toHaveBeenCalledTimes(1);
                    expect(rejectMock).toHaveBeenCalledTimes(1);
                    expect(errorMock).toHaveBeenCalledTimes(1);
                    done();
                }, 10);
            })
        });

        test('should pass 1 execution when circuit is HALF_OPEN in 2 executions', async () => {
            breaker.halfOpen();
            const successMock = jest.fn();
            const rejectMock = jest.fn();
            breaker.event.on('success', successMock);
            breaker.event.on('reject', rejectMock);

            const promise = () => timeoutPromise(5);

            await Promise.allSettled([
                breaker.execute(promise),
                breaker.execute(promise),
                breaker.execute(promise),
            ]);

            await Promise.allSettled([
                breaker.execute(promise),
                breaker.execute(promise),
                breaker.execute(promise),
            ]);

            expect(successMock).toHaveBeenCalledTimes(4);
            expect(rejectMock).toHaveBeenCalledTimes(2);
        });

        test('should fail when passing a Promise instead of a function', async () => {
            const promise = successPromise('success') as any;
            await expect(breaker.execute(promise)).rejects.toThrow(
                'Invalid argument: Expected a function that returns a Promise. Ensure you are passing a function, not a Promise.'
            );
        });
    });

    describe('timeout', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({ timeout: 100 });
        });

        test('should execute successfully', async () => {
            const promise = () => successPromise('success');
            await expect(breaker.execute(promise)).resolves.toBe('success');
        });

        test('should throw error on timeout', async () => {
            const promise = () => timeoutPromise(200);
            await expect(breaker.execute(promise)).rejects.toThrow('Operation timed out');
        });

        test('should set state to OPEN on failure', async () => {
            const promise = () => failurePromise(10, new Error('failure'));
            await expect(breaker.execute(promise)).rejects.toThrow('failure');
            expect(breaker.isOpen()).toBe(true);
        });
    });

    describe('resetTimeout', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({ timeout: 10, resetTimeout: 50 });
        });

        test('shoud reset timeout after open', (done) => {
            breaker.open();
            expect(breaker.isOpen()).toBe(true);

            setTimeout(() => {
                expect(breaker.isHalfOpen()).toBe(true);
                done();
            }, 51);
        });

        test('shoud not reset timeout when resetTimeout is undefined', (done) => {
            breaker = new CircuitBreaker({ timeout: 10 });
            breaker.open();
            expect(breaker.isOpen()).toBe(true);

            setTimeout(() => {
                expect(breaker.isOpen()).toBe(true);
                done();
            }, 51);
        });
    });

    describe('state transitions', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({ timeout: 10, resetTimeout: 5 });
        });

        test('should transition from HALF_OPEN to CLOSE on successful request', async () => {
            breaker.halfOpen();
            expect(breaker.isHalfOpen()).toBe(true);
            const promise = () => successPromise('success');
            await expect(breaker.execute(promise)).resolves.toBe('success');
            expect(breaker.isClosed()).toBe(true);
        });

        test('should transition from HALF_OPEN to OPEN on failed request', async () => {
            breaker.halfOpen();
            expect(breaker.isHalfOpen()).toBe(true);
            const promise = () => failurePromise(5, new Error('failure'));
            await expect(breaker.execute(promise)).rejects.toThrow('failure');
            expect(breaker.isOpen()).toBe(true);
        });

        test('should reset state to HALF_OPEN after resetTimeout', async () => {
            breaker.open();
            await timeoutPromise(10);
            expect(breaker.isHalfOpen()).toBe(true);
        });
    });

    describe('state transitions with default options', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({});
        });

        test('should not throw timeout exeption when timeout is undefined', async () => {
            const promise = () => timeoutPromise(50);
            await breaker.execute(promise);
            expect(breaker.isClosed()).toBe(true);
        });
    });

    describe('isError', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({
                timeout: 100,
                resetTimeout: 5,
                isError: (err) => err.message !== 'non-critical error',
            });
        });

        test('should execute successfully', async () => {
            const promise = () => successPromise('success');
            await expect(breaker.execute(promise)).resolves.toBe('success');
        });

        test('should not consider non-critical error as a failure', async () => {
            const promise = () => failurePromise(5, new Error('non-critical error'));
            await expect(breaker.execute(promise)).rejects.toThrow('non-critical error');
            expect(breaker.isOpen()).toBe(false);
        });

        test('should consider critical error as a failure', async () => {
            const promise = () => failurePromise(5, new Error('critical error'));
            await expect(breaker.execute(promise)).rejects.toThrow('critical error');
            expect(breaker.isOpen()).toBe(true);
        });

        test('should throw error when circuit is open', async () => {
            breaker.open();
            const promise = () => successPromise('success');
            await expect(breaker.execute(promise)).rejects.toThrow('Circuit is open');
        });

        test('should re-throw error and open circuit for critical error', async () => {
            const errorEventMock = jest.fn();
            breaker.event.on('error', errorEventMock);

            const criticalError = new Error('critical error');
            const promise = () => failurePromise(5, criticalError);

            await expect(breaker.execute(promise)).rejects.toThrow('critical error');

            expect(errorEventMock).toHaveBeenCalledWith(criticalError);

            expect(breaker.isOpen()).toBe(true);
        });
    });

    describe('events', () => {
        let breaker: CircuitBreaker;
        let openListener: jest.Mock;
        let closeListener: jest.Mock;
        let halfOpenListener: jest.Mock;
        let successListener: jest.Mock;
        let errorListener: jest.Mock;

        beforeEach(() => {
            breaker = new CircuitBreaker({ timeout: 10, resetTimeout: 5 });
            openListener = jest.fn();
            closeListener = jest.fn();
            halfOpenListener = jest.fn();
            successListener = jest.fn();
            errorListener = jest.fn();

            breaker.event.on('open', openListener);
            breaker.event.on('close', closeListener);
            breaker.event.on('halfOpen', halfOpenListener);
            breaker.event.on('success', successListener);
            breaker.event.on('error', errorListener);
        });

        test('should emit open event', async () => {
            const promise = () => failurePromise(5, new Error('failure'));
            await expect(breaker.execute(promise)).rejects.toThrow('failure');
            expect(openListener).toHaveBeenCalled();
        });

        test('should emit close event', async () => {
            breaker.halfOpen();
            const promise = () => successPromise('success');
            await expect(breaker.execute(promise)).resolves.toBe('success');
            expect(closeListener).toHaveBeenCalled();
        });

        test('should emit halfOpen event', async () => {
            breaker.open();
            await timeoutPromise(11);
            expect(halfOpenListener).toHaveBeenCalled();
        });

        test('should emit success event', async () => {
            const promise = () => successPromise('success');
            await expect(breaker.execute(promise)).resolves.toBe('success');
            expect(successListener).toHaveBeenCalledWith('success');
        });

        test('should emit error event', async () => {
            const promise = () => failurePromise(5, new Error('failure'));
            await expect(breaker.execute(promise)).rejects.toThrow('failure');
            expect(errorListener).toHaveBeenCalledWith(expect.any(Error));
        });
    });

    describe('autoRenewAbortController', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({ timeout: 10, resetTimeout: 50, autoRenewAbortController: true });
        });

        test('shoud renew AbortController timeout after HALF_OPEN', (done) => {
            const oldSignal = breaker.signal;
            breaker.open();

            setTimeout(async () => {
                expect(breaker.isHalfOpen()).toBe(true);
                const newSignal = breaker.signal;
                expect(oldSignal?.aborted).toBe(true);
                expect(newSignal?.aborted).toBe(false);
                done();
            }, 51);
        });

        test('shoud renew AbortController timeout after OPEN', (done) => {
            const oldSignal = breaker.signal;
            breaker.open();

            setTimeout(async () => {
                const promise = () => successPromise('success');
                await expect(breaker.execute(promise)).resolves.toBe('success');
                expect(breaker.isClosed()).toBe(true);
                const newSignal = breaker.signal;
                expect(oldSignal?.aborted).toBe(true);
                expect(newSignal?.aborted).toBe(false);
                done();
            }, 51);
        });
    });

    describe('failureThresholdCount', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({ timeout: 10, resetTimeout: 100, failureThresholdCount: 3 });
        });

        test('should set state to OPEN on max failureThresholdCount', async () => {
            const promise = () => successPromise('success');
            const failPromise = () => Promise.reject('failure');

            await Promise.all(Array.from({ length: 100 }).map(() => breaker.execute(promise)));
            await Promise.allSettled([
                breaker.execute(failPromise),
                breaker.execute(failPromise),
            ]);

            expect(breaker.isClosed()).toBe(true);
        });

        test('should set state to OPEN on max failureThresholdCount', async () => {
            const promise = () => successPromise('success');
            const failPromise = () => Promise.reject('failure');

            await Promise.all(Array.from({ length: 100 }).map(() => breaker.execute(promise)));
            await Promise.allSettled([
                breaker.execute(failPromise),
                breaker.execute(failPromise),
                breaker.execute(failPromise),
            ]);

            expect(breaker.isOpen()).toBe(true);
        });

        test('should set state to OPEN passing max failureThresholdCount', async () => {
            const promise = () => successPromise('success');
            const failPromise = () => Promise.reject('failure');

            await Promise.all(Array.from({ length: 100 }).map(() => breaker.execute(promise)));
            await Promise.allSettled([
                breaker.execute(failPromise),
                breaker.execute(failPromise),
                breaker.execute(failPromise),
                breaker.execute(failPromise),
            ]);

            expect(breaker.isOpen()).toBe(true);
        });
    });

    describe('successThreshold', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({ timeout: 10, resetTimeout: 100, successThreshold: 2 });
        });

        test('should remain in HALF_OPEN state when successThreshold is below limit', async () => {
            const promise = () => successPromise('success');
            breaker.halfOpen();

            await breaker.execute(promise);

            expect(breaker.isHalfOpen()).toBe(true);
        });

        test('should transition to CLOSED state when reaching successThreshold limit', async () => {
            const promise = () => successPromise('success');
            breaker.halfOpen();

            await breaker.execute(promise);
            await breaker.execute(promise);

            expect(breaker.isClosed()).toBe(true);
        });

        test('should transition to CLOSED state when exceeding successThreshold limit', async () => {
            const promise = () => successPromise('success');
            breaker.halfOpen();

            await breaker.execute(promise);
            await breaker.execute(promise);
            await breaker.execute(promise);

            expect(breaker.isClosed()).toBe(true);
        });

        test('should remain HALF_OPEN state after reset and one successful attempt', async () => {
            const promise = () => successPromise('success');

            breaker.halfOpen();
            await breaker.execute(promise);
            await breaker.execute(promise);

            breaker.halfOpen();
            await breaker.execute(promise);

            expect(breaker.isHalfOpen()).toBe(true);
        });

        test('should transition to CLOSED state after reset and reaching successThreshold', async () => {
            const promise = () => successPromise('success');

            breaker.halfOpen();
            await breaker.execute(promise);
            await breaker.execute(promise);

            breaker.halfOpen();
            await breaker.execute(promise);
            await breaker.execute(promise);

            expect(breaker.isClosed()).toBe(true);
        });
    });

    describe('retryAttempts', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({ timeout: 1, retryAttempts: 2 });
        });

        test('should remain in HALF_OPEN state if retries are below attempt limit', async () => {
            const promise = () => failurePromise(5, new Error('failure'));

            breaker.halfOpen();
            await Promise.allSettled([breaker.execute(promise)]);

            expect(breaker.isHalfOpen()).toBe(true);
        });

        test('should transition to OPEN state upon reaching retry attempt limit', async () => {
            const promise = () => failurePromise(5, new Error('failure'));

            breaker.halfOpen();
            await Promise.allSettled([breaker.execute(promise)]);
            await Promise.allSettled([breaker.execute(promise)]);

            expect(breaker.isOpen()).toBe(true);
        });

        test('should transition to OPEN state upon exceeding retry attempt limit', async () => {
            const promise = () => failurePromise(5, new Error('failure'));

            breaker.halfOpen();
            await Promise.allSettled([breaker.execute(promise)]);
            await Promise.allSettled([breaker.execute(promise)]);
            await Promise.allSettled([breaker.execute(promise)]);

            expect(breaker.isOpen()).toBe(true);
        });

        test('should remain HALF_OPEN after reset and one failed retry', async () => {
            const promise = () => failurePromise(5, new Error('failure'));

            breaker.halfOpen();
            await Promise.allSettled([breaker.execute(promise)]);
            await Promise.allSettled([breaker.execute(promise)]);

            breaker.halfOpen();
            await Promise.allSettled([breaker.execute(promise)]);

            expect(breaker.isHalfOpen()).toBe(true);
        });

        test('should transition to OPEN state after reset and retry limit is reached', async () => {
            const promise = () => failurePromise(5, new Error('failure'));

            breaker.halfOpen();
            await Promise.allSettled([breaker.execute(promise)]);
            await Promise.allSettled([breaker.execute(promise)]);

            breaker.halfOpen();
            await Promise.allSettled([breaker.execute(promise)]);
            await Promise.allSettled([breaker.execute(promise)]);

            expect(breaker.isOpen()).toBe(true);
        });
    });

    describe('toJSON', () => {
        let breaker: CircuitBreaker;

        beforeEach(() => {
            breaker = new CircuitBreaker({});
        });

        test('should match snapshot to default options state', async () => {
            expect(breaker.toJSON()).toMatchSnapshot();
        });

    });
});
