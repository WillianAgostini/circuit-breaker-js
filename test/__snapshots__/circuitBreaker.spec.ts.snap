// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircuitBreaker toJSON should match snapshot to default options state 1`] = `
{
  "failureCount": 0,
  "failurePercentage": 0,
  "isClosed": true,
  "isHalfOpen": false,
  "isOpen": false,
  "options": BreakerOptions {
    "autoRenewAbortController": false,
    "failureThresholdCount": 0,
    "failureThresholdPercentage": 5,
    "isError": undefined,
    "resetTimeout": undefined,
    "retryAttempts": 1,
    "successThreshold": 1,
    "timeout": undefined,
    "windowSize": 60000,
  },
  "successCount": 0,
  "totalRequests": 0,
}
`;
